tree.go

```diff
@@ -683,62 +685,86 @@ func (n *InternalNode) Commit() *Point {
 	internalNodeLevels := make([][]*InternalNode, StemSize)
 	n.fillLevels(internalNodeLevels)
 
-	points := make([]*Point, 0, 1024)
-	cowIndexes := make([]int, 0, 1024)
-	poly := make([]Fr, NodeWidth)
 	for level := len(internalNodeLevels) - 1; level >= 0; level-- {
 		nodes := internalNodeLevels[level]
 		if len(nodes) == 0 {
 			continue
 		}
-		points = points[:0]
-		cowIndexes = cowIndexes[:0]
-
-		// For each internal node, we collect in `points` all the ones we need to map to a field element.
-		// That is, for each touched children in a node, we collect the old and new commitment to do the diff updating
-		// later.
-		for _, node := range nodes {
-			for idx, nodeChildComm := range node.cow {
-				points = append(points, nodeChildComm)
-				points = append(points, node.children[idx].Commitment())
-				cowIndexes = append(cowIndexes, int(idx))
+
+		minBatchSize := 4
+		if len(nodes) <= minBatchSize {
+			commitNodesAtLevel(nodes)
+		} else {
+			var wg sync.WaitGroup
+			numBatches := runtime.NumCPU()
+			batchSize := (len(nodes) + numBatches - 1) / numBatches
+			if batchSize < minBatchSize {
+				batchSize = minBatchSize
+			}
+			for i := 0; i < len(nodes); i += batchSize {
+				start := i
+				end := i + batchSize
+				if end > len(nodes) {
+					end = len(nodes)
+				}
+				wg.Add(1)
+				go func() {
+					defer wg.Done()
+					commitNodesAtLevel(nodes[start:end])
+				}()
```

Comment:
TL;DR: we divide the list of nodes to use at most all available cores, but each core at least doing minimum amount of work.